### ISSUE TITLE
Add --skip-auth flag (or similar) to support automated deployments

### ISSUE THREAD

**Issue by TitanChris** [certbot/certbot#2933]
Today, it seems very difficult to fully automate deployment to a new server, if using `certonly` and `--webroot`.

Consider the example of deploying nginx and Let's Encrypt certs to a new server.  The prevailing wisdom online is that you:
- (A1) Make sure nginx is running, with Config X1 that enables webroot authentication but does NOT reference any cert files (which don't yet exist).
- (A2) Run `letsencrypt certonly ...` with webroot authentication.
- (A3) Change your nginx config file to Config X2, which is like Config X1 but also references the generated certs. [You cannot do step (A3) before (A2), because nginx won't start if non-existent cert files are referenced.] [And no, I don't want to trust the Let's Encrypt nginx plugin to get this right on production servers. :)]

Unfortunately, this procedure breaks automated deployment.  Deploy rules say things like "ensure Config Y is in place".  Not to mention that even if you could, you really don't want to ping-pong between Config X1 and Config X2 -- and restart nginx twice -- on every subsequent deploy to that server.

As a solution, I would like to ask that Let's Encrypt support the following deploy procedure (in particular, step B1 below):
- (B1) Run `letsencrypt` in a way that does NO authentication, but DOES install dummy certs, for all domains specified, ONLY IF cert files do not yet exist.  (This is just to have files in place before the nginx config gets placed in step B2.)
- (B2) Make sure nginx is running, with Config X2, which references the generated certs and enables webroot authentication.
- (B3) Run `letsencrypt` in a way that DOES do authentication, and that installs real certs _IF_ there are currently dummy certs (or absent cert files), for all domains specified. (And _IF_ any real certs were installed, indicates that in some way, so that the deployment script can then restart nginx.)

Might it be as simple as adding a new `--skip-auth` flag, which could be combined with the existing `--test-cert` flag to enable step B1?

[As a temporary workaround today, I run an ugly custom script to do step B1. But that is silly, because (a) Let's Encrypt already knows which files will be created, in which paths, for which domains, and (b) duplicating hard-coded values between my custom script and my Let's Encrypt config is just asking for trouble.]

**Comment by pde** [comment]
Our next development priority is shipping the native nginx plugin, which should allow you to deploy a config with either dummy certs or no HTTPS support, then immediately upgrade them to real certs.

So this is probably a wontfix...

**Comment by pde** [comment]
You'll have the option of using `--nginx certonly`, which under the hood will get you certs by doing:

1. add new vhosts to your running nginx config to pass tls-sni challenges
2. soft reload conf
3. passing authentication / obtain certs
4. revert nginx conf to exact prior state
5. soft reload conf

Or you'll be able to run without `certonly` in which case the following extra steps will occur:

6. modify config to use new certs
7. soft reload, check for detectable nginx conf errors
8. if errors are detected, revert config

**Comment by TitanChris** [comment]
Those both sound like useful features. However, I don't think they solve the issue of automated deployments (e.g., via SaltStack or similar) that I described.

In the simpler scenario, where you tell the LE client to modify your nginx config, automated deploys still break.  During the next deploy to the same server, the deploy process will overwrite the (unrecognized, not-known-good) modified config with a deployment-managed config.

In the other scenario, where you tell the LE client to modify-and-revert your nginx config, you still have a chicken-and-egg problem. Nginx needs to be running, which means you need to have deployed a config. But ping-ponging between configs on every push is bad news, so the config already needs to reference the SSL cert file locations. But those files don't exist yet because the LE client hasn't run. But to generate those files nginx needs to be running....

**Comment by TitanChris** [comment]
My original report sort of buried the feature request, so let me try to state it more clearly:

  * **The LE client should allow writing dummy, self-signed certs / keys.**
  * Details:
    * It should use all the usual args (e.g., `--domain` and `--config`) to determine where to write these files.
    * It should NOT overwrite valid certs that already exist at these file locations.
    * It should NOT expect a webserver to be running.
    * It must write correctly formatted certs / keys (e.g., not just `touch` the files), otherwise nginx will refuse to load.

And the purpose of this mode is to support automated deployment, so that deploy rules can:
  1. Run the LE client to write the dummy certs (if none exist).
  1. Place a deploy-managed nginx config file that references these certs.
  1. Ensure nginx is running.
  1. Run the LE client to generate real certs, overwriting the dummy certs.

**Comment by md2k** [comment]
What about scenario when i have several servers, it is mean i should deploy bot on each server, ensure then each virtual domain able server data from /.well-known/* and each server should maintain its own certificate with subset of domains or cert per domain, i found this a bit painful.
Maybe it will be much easier to allow some central point of authorisation, when for example we have:
auth.domain.com hold .well-known folder where certbot place file for authorisation, this can help fully automate cert creation and renewal even in multi-server setups.
Or as an option some type of authorisation by API key / credentials.
Formally if some one will get access to server where is certbot installed any type of authentication will be compromised and certificates will be stolen, so user/pass or API Key way also not bad way to do that.

**Comment by pde** [comment]
Hi @TitanChris.  Using `--test-cert` / `--staging` flag will produce non-valid certs (they're issued by a staging version of the Let's Encrypt API).  It will never overwrite existing valid certs (unless you provide `--break-my-certs`).  What you probably want is that flag, with the `webroot` plugin.

There is also a `--dry-run` flag which implies `--staging`, but tries to basically test all the steps of obtaining a cert without actually writing it to disk.  It only supports `certonly` at the moment, but we should add support for `run` too.

**Comment by pde** [comment]
@md2k one good way to centralize authorization is to serve redirects from `.well-known/acme-challenge/` to your authorization machine, which can then obtain the certs and provide or push them out to each of your boxes.

**Comment by mijamo** [comment]
I ran into this issue and would like very much an option to create dummy / self-signed certificates for domain through a tag. We run our apps through Docker and try to have as much as possible the same settings in production and development, so right now our only solution with Certbot is to have a Docker container that creates self-signed certificates  first and creates real one when asked to through the `run`command. Unfortunately generating the chain is not funny at all and it is not a nice solution in any way. The nginx plugin does not help here because Certbot runs in a different container.

`--test-cert`does not help either in that case obviously because ideally that should be able to work disconnected from internet basically, or at least without any DNS record pointing to the machine.

**Comment by TitanChris** [comment]
@mijamo - I found that the [simp_le client](https://github.com/kuba/simp_le) worked better for automating my deployment. Using it, I was able to avoid the bootstrapping problem described in this thread. You may have similar luck.

**Comment by davidjb** [comment]
I'm seeing the same issue -- chicken-and-the egg problem with automated deployments.  I can't start `certbot --nginx` because when it calls `nginx` it is looking for certbot's certificates.  So either I'd need to temporarily remove the HTTPS part of the config and put it back in later (awkward) or do what I've done so far which is to generate a self-signed cert in the same location as the future live certs.  `nginx` is thus able to start and so `certbot --nginx` can proceed, overwriting the certs.

In short, @TitanChris's suggestion from above; detect in the nginx config if the corresponding `server {}` block for the given hostname is pointing at `certbot` certs and if non-existent, create self-signed certs temporarily so nginx can start.

**Comment by bmw** [comment]
I do see value in a feature like this, but to be honest, I doubt the small Certbot team has the resources to get to it anytime soon.

I think we'd take a well written PR for this feature though. If someone is interested in taking this on, I'd encourage you to propose a design here first and get a +1 from someone on the Certbot team before diving too deeply into writing code and tests.
